File: Algorithmic_Polytech/1-Complexite/code/puissance.c
#include <stdio.h>
#include <stdlib.h>

double puissance(double a, int n) {
    int i;
    double r = 1;
    for(i = 0; i < n; i++) {
        r = r*a;
    }
    return r;
}

int main(int argc, char *argv[]) {
    double a = 1;
    int n = 1;

    if(argc > 1) a = atof(argv[1]);
    if(argc > 2) n = atoi(argv[2]);

    printf("%f\n",puissance(a,n));

    return 0;

}

File: Algorithmic_Polytech/1-Complexite/code/puissance.py
import os, sys

def puissance(a,n):
    d = n.bit_length()-1
    r = 1
    while d >= 0:
        r = r*r
        if 1 & (n >> d) == 1:
            r = r*a
        d -= 1
    return r

a = 1
n = 1

if len(sys.argv) > 1:
    a = float(sys.argv[1])

if len(sys.argv) > 2:
    n = int(sys.argv[2])

print(puissance(a,n))
